### PR TITLE
Fix: Keep rules sorted

### PR DESCRIPTION
**Pull request by localheinz**
This PR

* [x] keeps rules sorted in `.php_cs.dist`

### CODE REVIEW

**Comment by SpacePossum** [comment]
could we auto review this by unit test you think?

**Comment by localheinz** [comment]
@SpacePossum 

Only if we extracted an internal ruleset that we could then inspect, I believe.

**Comment by SpacePossum** [comment]
```php
<?php

require_once __DIR__.'/../vendor/autoload.php';

// assert file exists

/** @var \PhpCsFixer\Config $config */
$config = include __DIR__.'/../.php_cs.dist';

// assert instance of Config

foreach ($config->getRules() as $rule => $config) {
    echo sprintf("\n%s", $rule); // assert `@` first, all `@[n]` are sorted, all rules (non-sets) are sorted
}

die;
```

List the rules `as is` as these are not resolved yet, so we could check I think.
We might need the `ConfigurationResolver::separatedContextLessInclude` trick to prevent from including the configuration only one time (we might need the multiple times in the tests).

**Comment by keradus** [comment]
OK, I will drop RTM to await for tests ;)
`separatedContextLessInclude` has nothing to do with preventing from including multiple times, it's only about not sharing context

**Comment by SpacePossum** [comment]
```
2) PhpCsFixer\Tests\AutoReview\ProjectFixerConfigurationTest::testRuleDefinedAlpha
--- Original
+++ New
@@ @@
   33 => 'no_useless_return',
   34 => 'ordered_class_elements',
   35 => 'ordered_imports',
-  41 => 'phpdoc_add_missing_param_annotation',
-  42 => 'phpdoc_order',
-  43 => 'phpdoc_types_order',
   36 => 'php_unit_ordered_covers',
   37 => 'php_unit_set_up_tear_down_visibility',
   38 => 'php_unit_strict',
   39 => 'php_unit_test_annotation',
   40 => 'php_unit_test_class_requires_covers',
+  41 => 'phpdoc_add_missing_param_annotation',
+  42 => 'phpdoc_order',
+  43 => 'phpdoc_types_order',
   44 => 'semicolon_after_instruction',
   45 => 'single_line_comment_style',
   46 => 'strict_comparison',
```

we want this or not `natcasesort` but just `sort`?

**Comment by keradus** [comment]
i would use normal sort, as we are using normal sort everywhere else.
but no big deal here

File: tests/AutoReview/ProjectFixerConfigurationTest.php [hunk 1]
```diff
@@ -26,23 +28,46 @@
...
+
+    public function testRuleDefinedAlpha()
+    {
+        $rules = $rulesSorted = array_keys($this->config->getRules());
+        natcasesort($rulesSorted);
+
+        $this->assertSame(
+            $rulesSorted,
+            $rules,
+            (new Differ())->diff(var_export($rulesSorted, true), var_export($rules, true))
```
Comment (keradus) [inline tests/AutoReview/ProjectFixerConfigurationTest.php]: isn't failing assertion provides the diff already ?

Comment (SpacePossum) [reply tests/AutoReview/ProjectFixerConfigurationTest.php]: it kind of does (ignore the first failure);
```
$ ./vendor/bin/phpunit tests/AutoReview/ProjectFixerConfigurationTest.php
PHPUnit 6.5.7 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.3-1+ubuntu16.04.1+deb.sury.org+1
Configuration: /home/possum/work/PHP-CS-Fixer/phpunit.xml.dist

Testing PhpCsFixer\Tests\AutoReview\ProjectFixerConfigurationTest
FF                                                                                                                                                                  2 / 2 (100%)

Time: 112 ms, Memory: 10.00MB

There were 2 failures:

1) PhpCsFixer\Tests\AutoReview\ProjectFixerConfigurationTest::testCreate
Failed asserting that actual size 151 matches expected size 50.

/home/possum/work/PHP-CS-Fixer/tests/AutoReview/ProjectFixerConfigurationTest.php:59

2) PhpCsFixer\Tests\AutoReview\ProjectFixerConfigurationTest::testRuleDefinedAlpha
Failed asserting that Array &0 (
    0 => '@PHP56Migration'
    1 => '@PHPUnit60Migration:risky'
    2 => '@Symfony'
    3 => '@Symfony:risky'
    4 => 'align_multiline_comment'
    5 => 'array_indentation'
    6 => 'array_syntax'
    7 => 'blank_line_before_statement'
    8 => 'combine_consecutive_issets'
    9 => 'combine_consecutive_unsets'
    10 => 'comment_to_phpdoc'
    11 => 'compact_nullable_typehint'
    12 => 'escape_implicit_backslashes'
    13 => 'explicit_indirect_variable'
    14 => 'explicit_string_variable'
    15 => 'final_internal_class'
    16 => 'fully_qualified_strict_types'
    17 => 'function_to_constant'
    18 => 'header_comment'
    19 => 'heredoc_to_nowdoc'
    20 => 'list_syntax'
    21 => 'method_argument_space'
    22 => 'method_chaining_indentation'
    23 => 'multiline_comment_opening_closing'
    24 => 'no_alternative_syntax'
    25 => 'no_extra_blank_lines'
    26 => 'no_null_property_initialization'
    27 => 'no_short_echo_tag'
    28 => 'no_superfluous_elseif'
    29 => 'no_unneeded_curly_braces'
    30 => 'no_unneeded_final_method'
    31 => 'no_unreachable_default_argument_value'
    32 => 'no_useless_else'
    33 => 'no_useless_return'
    34 => 'ordered_class_elements'
    35 => 'ordered_imports'
    36 => 'php_unit_ordered_covers'
    37 => 'php_unit_set_up_tear_down_visibility'
    38 => 'php_unit_strict'
    39 => 'php_unit_test_annotation'
    40 => 'php_unit_test_class_requires_covers'
    41 => 'phpdoc_add_missing_param_annotation'
    42 => 'phpdoc_order'
    43 => 'phpdoc_types_order'
    44 => 'semicolon_after_instruction'
    45 => 'single_line_comment_style'
    46 => 'strict_comparison'
    47 => 'strict_param'
    48 => 'string_line_ending'
    49 => 'yoda_style'
) is identical to Array &0 (
    0 => '@PHP56Migration'
    1 => '@PHPUnit60Migration:risky'
    2 => '@Symfony'
    3 => '@Symfony:risky'
    4 => 'align_multiline_comment'
    5 => 'array_indentation'
    6 => 'array_syntax'
    7 => 'blank_line_before_statement'
    8 => 'combine_consecutive_issets'
    9 => 'combine_consecutive_unsets'
    10 => 'comment_to_phpdoc'
    11 => 'compact_nullable_typehint'
    12 => 'escape_implicit_backslashes'
    13 => 'explicit_indirect_variable'
    14 => 'explicit_string_variable'
    15 => 'final_internal_class'
    16 => 'fully_qualified_strict_types'
    17 => 'function_to_constant'
    18 => 'header_comment'
    19 => 'heredoc_to_nowdoc'
    20 => 'list_syntax'
    21 => 'method_argument_space'
    22 => 'method_chaining_indentation'
    23 => 'multiline_comment_opening_closing'
    24 => 'no_alternative_syntax'
    25 => 'no_extra_blank_lines'
    26 => 'no_null_property_initialization'
    27 => 'no_short_echo_tag'
    28 => 'no_superfluous_elseif'
    29 => 'no_unneeded_curly_braces'
    30 => 'no_unneeded_final_method'
    31 => 'no_unreachable_default_argument_value'
    32 => 'no_useless_else'
    33 => 'no_useless_return'
    34 => 'ordered_class_elements'
    35 => 'ordered_imports'
    41 => 'phpdoc_add_missing_param_annotation'
    42 => 'phpdoc_order'
    43 => 'phpdoc_types_order'
    36 => 'php_unit_ordered_covers'
    37 => 'php_unit_set_up_tear_down_visibility'
    38 => 'php_unit_strict'
    39 => 'php_unit_test_annotation'
    40 => 'php_unit_test_class_requires_covers'
    44 => 'semicolon_after_instruction'
    45 => 'single_line_comment_style'
    46 => 'strict_comparison'
    47 => 'strict_param'
    48 => 'string_line_ending'
    49 => 'yoda_style'
).

/home/possum/work/PHP-CS-Fixer/tests/AutoReview/ProjectFixerConfigurationTest.php:69
```

that diff is kind of hard to read, we can drop the Differ when this PR is close to finished

Comment (keradus) [reply tests/AutoReview/ProjectFixerConfigurationTest.php]: ok, let us keep it

File: tests/AutoReview/ProjectFixerConfigurationTest.php [hunk 2]
```diff
@@ -26,23 +28,44 @@
...
+
+    public function testRuleDefinedAlpha()
+    {
+        $rules = $rulesSorted = array_keys($this->config->getRules());
+        sort($rulesSorted);
+
+        $this->assertSame(
+            $rulesSorted,
+            $rules,
+            (new Differ())->diff(var_export($rulesSorted, true), var_export($rules, true))
```
Comment (keradus) [inline tests/AutoReview/ProjectFixerConfigurationTest.php]: why `Diff` and not `Differ` ?

Comment (SpacePossum) [reply tests/AutoReview/ProjectFixerConfigurationTest.php]: not sure what you mean?

**Comment by keradus** [comment]
well... 2.2 ?

**Comment by SpacePossum** [comment]
@localheinz if you agree, please squash my doodles-commits into your first commit and do the retarget to 2.2, I don't want to mess around even more with your branch than I already did

**Comment by keradus** [comment]
on 2.2 there won't be changes in `.php_cs` files, eg `no_alternative_syntax` were not existing there.
@SpacePossum , simply open new PR of your commits only targetting 2.2, it will replace this PR

**Comment by SpacePossum** [comment]
check, done, @localheinz please throw out my commits here (sorry for the noise)

**Comment by keradus** [comment]
replaced by #3629
Thanks @localheinz !

**Pull request closed without merging**